src/runtime-rs/crates/hypervisor/src/device/driver/mod.rs

```diff
@@ -4,20 +4,188 @@
+mod virtio_fs;
+mod virtio_net;
+mod virtio_vsock;
+
+pub use vfio::{
+    bind_device_to_host, bind_device_to_vfio, get_vfio_device, HostDevice, VfioBusMode, VfioConfig,
+    VfioDevice,
+};
 pub use virtio_blk::{
     BlockConfig, BlockDevice, KATA_BLK_DEV_TYPE, KATA_MMIO_BLK_DEV_TYPE, VIRTIO_BLOCK_MMIO,
     VIRTIO_BLOCK_PCI,
 };
-mod virtio_net;
-pub use virtio_net::{Address, NetworkConfig, NetworkDevice};
-mod vfio;
-pub use vfio::{bind_device_to_host, bind_device_to_vfio, VfioBusMode, VfioConfig, VfioDevice};
-mod virtio_fs;
 pub use virtio_fs::{
     ShareFsDevice, ShareFsDeviceConfig, ShareFsMountConfig, ShareFsMountDevice, ShareFsMountType,
     ShareFsOperation,
 };
-mod virtio_vsock;
+pub use virtio_net::{Address, NetworkConfig, NetworkDevice};
 pub use virtio_vsock::{HybridVsockConfig, HybridVsockDevice, VsockConfig, VsockDevice};
+
+// The PCI spec reserves 5 bits for slot number (a.k.a. device
+// number), giving slots 0..31
+const PCI_SLOT_BITS: u32 = 5;
+const MAX_PCI_SLOTS: u32 = (1 << PCI_SLOT_BITS) - 1;
+
+// A PciSlot describes where a PCI device sits on a single bus
+//
+// This encapsulates the PCI slot number a.k.a device number, which is
+// limited to a 5 bit value [0x00..0x1f] by the PCI specification
+//
+// In order to support multifunction device's we'll need to extend
+// this to include the PCI 3-bit function number as well.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct PciSlot(pub u8);
+
+impl PciSlot {
+    pub fn pci_slot_from_string(s: &str) -> Option<PciSlot> {
+        if s.is_empty() || s.len() > 2 {
+            return None;
+        }
+
+        let base = 16;
+        match u64::from_str_radix(s, base) {
+            // exceed 31 return None
+            Ok(n) if n >> PCI_SLOT_BITS == 0 => Some(PciSlot(n as u8)),
+            _ => None,
+        }
+    }
+
+    pub fn pci_slot_from_u32(v: u32) -> PciSlot {
+        let slot = if v > MAX_PCI_SLOTS { MAX_PCI_SLOTS } else { v };
+
+        PciSlot(slot as u8)
+    }
+
+    pub fn pci_slot_to_string(&self) -> String {
+        format!("{:02x}", self.0)
+    }
+}
+
+// A PciPath describes where a PCI sits in a PCI hierarchy.
+//
+// Consists of a list of PCI slots, giving the slot of each bridge
+// that must be traversed from the PCI root to reach the device,
+// followed by the slot of the device itself.
+//
+// When formatted into a string is written as "xx/.../yy/zz". Here,
+// zz is the slot of the device on its PCI bridge, yy is the slot of
+// the bridge on its parent bridge and so forth until xx is the slot
+// of the "most upstream" bridge on the root bus.
+//
+// If a device is directly connected to the root bus, which used in
+// lightweight hypervisors, such as dragonball/firecracker/clh, and
+// its PciPath.slots will contains only one PciSlot.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct PciPath {
+    // list of PCI slots
+    slots: Vec<PciSlot>,
+}
+
+impl PciPath {
+    // method to format the PciPath into a string
```

Comment:
Suggest removing the prefix `pci_path` for function in `PciPath`.

Reply:
fixed